Source/WebCore/platform/graphics/mac/controls/ButtonMac.mm

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All Rights Reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+#import "ButtonMac.h"
+
+#if PLATFORM(MAC)
+
+#import "ButtonPart.h"
+#import "ControlFactoryMac.h"
+#import "GraphicsContext.h"
+#import "LocalDefaultSystemAppearance.h"
+#import <wtf/BlockObjCExceptions.h>
+
+namespace WebCore {
+
+ButtonMac::ButtonMac(ButtonPart& owningPart, ControlFactoryMac& controlFactory, NSButtonCell *buttonCell)
+    : ButtonControlMac(owningPart, controlFactory, buttonCell)
+{
+    ASSERT(m_owningPart.type() == ControlPartType::Button
+        || m_owningPart.type() == ControlPartType::DefaultButton
+        || m_owningPart.type() == ControlPartType::PushButton
+        || m_owningPart.type() == ControlPartType::SquareButton);
+}
+
+IntSize ButtonMac::cellSize(NSControlSize controlSize, const ControlStyle&) const
+{
+    // Buttons really only constrain height. They respect width.
+    static const IntSize cellSizes[] = {
+        { 0, 21 },
+        { 0, 18 },
+        { 0, 15 },
+        { 0, 28 }
+    };
+    return cellSizes[controlSize];
+}
+
+IntOutsets ButtonMac::cellOutsets(NSControlSize controlSize, const ControlStyle&) const
+{
+    // FIXME: These values may need to be reevaluated. They appear to have been originally chosen
+    // to reflect the size of shadows around native form controls on macOS, but as of macOS 10.15,
+    // these margins extend well past the boundaries of a native button cell's shadows.
+    static const IntOutsets cellOutsets[] = {
+        // top right bottom left
+        { 4, 6, 7, 6 },
```

Comment:
Is there a plan to remove these values from `ThemeMac`?

Reply:
Yes they will be removed once all the references to buttonSizes() and buttonMargins() are removed. I will fix this after refactoring the painting code for all the controls.